File: gatewayapi.spring/src/main/java/com/example/gatewayapi/controller/AccountController.java
package com.example.gatewayapi.controller;

import com.example.account.model.AccountModel;
import com.example.common.rest.dto.ErrorDto;
import com.example.common.rest.envelope.ResponseEnvelope;
import com.example.gatewayapi.exception.ExceptionCode;
import com.example.gatewayapi.exception.GatewayApiException;
import com.example.gatewayapi.service.AccountService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.List;

@Controller
@RequestMapping(value = "/api/v1/account", produces = MediaType.APPLICATION_JSON_VALUE)
public class AccountController extends AbstractController {

    private static final Logger log = LoggerFactory.getLogger(AccountController.class);

    private AccountService accountService;

    @Autowired
    public AccountController(AccountService accountService) {
        this.accountService = accountService;
    }

    @RequestMapping(value = "/{accountId}", method = RequestMethod.GET)
    public ResponseEntity getAccountDetails(
            @RequestHeader("authkey") String token,
            @PathVariable("accountId") BigInteger accountId
    ) {
        //todo validate input
        List<ErrorDto> errors = new ArrayList<>();
        HttpStatus responseStatus = HttpStatus.OK;

        AccountModel response = null;
        try {
            String tenantId = getTenantId(token);
            response = accountService.getAccount(tenantId, accountId);
        } catch (GatewayApiException e) {
            log.error(e.getMessage(), e);
            errors.add(new ErrorDto(ExceptionCode.ACCOUNT_NOT_FOUND.name(), ExceptionCode.ACCOUNT_NOT_FOUND.getMessage()));
            responseStatus = HttpStatus.NOT_FOUND;
        }

        ResponseEnvelope responseEnvelope = new ResponseEnvelope.Builder<AccountModel>()
                .withData(response)
                .withErrors(errors)
                .build();

        return ResponseEntity
                .status(responseStatus)
                .body(responseEnvelope);
    }

}

File: services/account/account-api/src/main/kotlin/com/example/account/api/rest/controller/AccountController.kt
package com.example.account.api.rest.controller

import com.example.account.api.facade.AccountFacade
import com.example.account.api.rest.controller.AccountController.Companion.PROTOBUF_MEDIA_TYPE_VALUE
import com.example.account.api.rest.converter.AccountModelConverter
import com.example.account.api.rest.exception.AccountException
import com.example.account.api.rest.exception.ExceptionCode
import com.example.account.api.rest.validator.RequestValidator
import com.example.account.api.spec.model.AAccountDto
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping(value = ["/api/v1"], produces = [PROTOBUF_MEDIA_TYPE_VALUE, MediaType.APPLICATION_JSON_VALUE])
class AccountController(private val accountFacade: AccountFacade) {

    companion object {
        const val PROTOBUF_MEDIA_TYPE_VALUE = "application/x-protobuf"
        private val LOG = LoggerFactory.getLogger(AccountController::class.java)
    }


    @RequestMapping(value = ["", "/"], method = [RequestMethod.GET])
    @ResponseStatus(HttpStatus.OK)
    fun getAccountByUserAndPassword(
            @RequestParam(name = "username", defaultValue = "") username: String,
            @RequestParam(name = "password", defaultValue = "") password: String
    ): AAccountDto {
        if (!RequestValidator.validateGetAccountByUserAndPassword(username, password)) {
            throw AccountException(ExceptionCode.BAD_REQUEST)
        }
        val accountModel = accountFacade.getAccount(username, password)

        if (accountModel == null || accountModel.isEmpty) {
            throw AccountException(ExceptionCode.ACCOUNT_NOT_FOUND)
        }
        return AccountModelConverter.toDto(accountModel.get())
    }
}